:bug: Fix #496 - Option buttons shouldn't move away from valid match.

colllin opened the pull request:
This is a proposal to fix #496.

The problem
---
Toggling the option buttons will jump you away from a selection that already matched the new settings.

The previous behavior
---
Always jump to the next match after the cursor whenever a find option is toggled.

The new behavior
---
Only jump to the next match **if none of the existing selections match** the pattern.  Inversely, **if any of the existing selections match** the pattern, it will not jump you to the next match.

Alternatives
---
I also considered and decided against these possible behaviors:

- Jump to the next match **if any of the existing selections don't match** the pattern.  Inversely, do not jump to the next match **if all of the existing selections match** the pattern.
- When a find option is toggled, destroy any selections that don't match.  If there are any remaining selections, don't jump to the next match.  Otherwise jump to the next match.

colllin commented:
Screen captures...

Before: always jumps even if existing selection already matched.
![before](https://cloud.githubusercontent.com/assets/185923/11314140/0efb86bc-8fa0-11e5-9374-33f974dd8a45.gif)

After: doesn't jump if existing selection matches.
![after-nojump](https://cloud.githubusercontent.com/assets/185923/11314150/2ade7876-8fa0-11e5-9103-2108355e3c4e.gif)

After: jumps if existing selection doesn't match.
![after-jump](https://cloud.githubusercontent.com/assets/185923/11314159/47d00134-8fa0-11e5-9369-109ef77e478b.gif)

50Wliu commented:
I like this.  Refs #298.

benogle commented on lib/find-view.coffee:
```diff
@@ -447,21 +447,30 @@ class FindView extends View
+    selections = editor?.getSelections() or []
+    _.any selections, (selection) =>
+      selectionMatchesFindPattern = false
+      editor.scanInBufferRange @model.getFindPatternRegex(), selection.getBufferRange(), ({range}) ->
```
This scan seems unnecessary. The model should know if there is a current result marker IIRC. Seems like we should just be able make the decision based on the existence of the current result marker.

benogle commented:
I'm down with this functionality. Need someone to look over all the code still though.

benogle commented on lib/find-view.coffee:
```diff
@@ -447,21 +447,27 @@ class FindView extends View
+    selections = @model.getEditor()?.getSelections() or []
+    _.any selections, (selection) =>
+      _.any @model.markers or [], (marker) =>
+        _.isEqual(marker.getBufferRange(), selection.getBufferRange())
```
I'm not keen on the underscore usage here. If there are a ton of markers this will probably break down. @maxbrunsfeld  @nathansobo is there a nice way to check if any markers are selected?

maxbrunsfeld replied:
@colllin I think that we need to avoid a `selections.length x markers.length` loop here, mainly because there can be hundreds of thousands of markers. I think the performance will be more predictable if you query for the markers by range, using `TextEditor::findMarkers`. There is already code to do this in [`BufferSearch.prototype.findMarker`](https://github.com/atom/find-and-replace/blob/d719c79b358cf59e15ca185091965d29a9ead49b/lib/buffer-search.coffee#L194). You could use that method like this:

```coffee
for selectedRange in @model.getEditor().getSelectedBufferRanges()
  if @model.findMarker(selectedRange)
    return true
```

maxbrunsfeld commented on spec/find-view-spec.coffee:
```diff
@@ -844,6 +844,37 @@ describe 'FindView', ->
 
+      describe "when there are existing selections", ->
+        describe "when any existing selections match the pattern", ->
+          it "shouldn't jump to the next match when toggled on", ->
```
Usually we avoid the `should` phrasing in the specs, opting for the slightly more direct `it "does not jump to the next match when toggled on"`.

maxbrunsfeld commented on spec/find-view-spec.coffee:
```diff
@@ -892,6 +956,39 @@ describe 'FindView', ->
+            editor.setSelectedBufferRange [[0, 0], [0, 5]]
+            findView.caseOptionButton.click()
+            expect(editor.getSelectedBufferRange()).toEqual [[1, 0], [1, 5]]
+
```
I usually don't say this, but I think there might be a few *too* many tests for this code change. They look very good, but I would probably combine the tests for `when toggled on` and `when toggled off` into one, in which you click the button twice, and check that the selection persists at every stage.

maxbrunsfeld commented:
Thanks for figuring this stuff out @colllin. I left a few suggested changes, but overall this looks great.

colllin commented:
Thank you for the suggestions and the tip about `Array.prototype.some`.

I...
- changed the `anyMarkersAreSelected` logic to use `BufferSearch.prototype.findMarker`
- combined the toggled on/off tests
- changed the spec phrasing to avoid "should"

maxbrunsfeld commented:
This is a great PR, @colllin. Thanks so much for sticking with it.

maxbrunsfeld commented:
This will be in Atom 1.4.0 and 1.4.0-beta0.
